In [1]:
import pandas as pd
import numpy as np
import datetime as dt
from IPython.display import clear_output
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import KFold, TimeSeriesSplit
from sklearn import neighbors


In [2]:
def find_nan_features(df):
    null_cols = []
    for col in df.columns:
        if df[col].isnull().values.any():
            null_cols.append(col)
    return null_cols

In [3]:
def remove_nan_rows(df):
    # getting indices (rows) of all NaN values
    inds = pd.isnull(df).any(1).nonzero()[0]

    # drop all the rows with NaN values
    return df.drop(df.index[inds])

In [4]:
def split(df, train_fraction):
    mindate = df.Date.min()
    maxdate = df.Date.max()
    splitdate = mindate + (maxdate - mindate) * train_fraction
    train = df[df.Date < splitdate]
    test = df[df.Date >= splitdate]
    return train, test

In [5]:
def get_x_y(df):
    # split set in data and target
    X = df.drop('NumberOfSales', axis=1)
    y = df["NumberOfSales"]
    return X, y

In [6]:
def train_model(X_train, y_train, n_estimators=250, n_jobs=1):
    # fit random forest with 250 trees
    model = neighbors.KNeighborsRegressor(n_neighbors=10, n_jobs=-1)
    model.fit(X_train, y_train)
    return model

In [7]:
def bip_metric(X_val, y_val, y_pred, rows_region):

    e_r = []
    month_sum = []
    # adjust shape
    X_val = X_val.reset_index(drop=True)
    X_val['Region'] = rows_region
    X_val = pd.get_dummies(X_val, columns=['Region'], prefix='Region')
    
    y_pred = y_pred.tolist()
    y_val = y_val.tolist()

    for r in range(11):

        region = 'Region_' + str(r)
        d = X_val.loc[X_val[region] == 1]

        error = 0
        y_somma = 0
  
#  cycle through stores
        for i in range(1000,1736): 
               
            for m in range(1,13):
                sum_pred_month = 0
                sum_actual_month = 0
                indexes = d.index[(d['StoreID'] == i) & (d['Month'] == m)].tolist()
                
                for j in indexes:
           
                    sum_pred_month += y_pred[j]
                    sum_actual_month += y_val[j]

                error += abs(sum_actual_month - sum_pred_month)
                y_somma = y_somma + sum_actual_month

    e_r.append(error/y_somma)

    return sum(e_r)/len(e_r)

In [8]:
def eval_model(X_val, y_val, model, store_ids, val_date, rows_region):
    y_pred = model.predict(X_val)
    new_x_val = X_val 
    new_x_val['Month'] = pd.DatetimeIndex(val_date['Date']).month
    new_x_val['StoreID'] = store_ids
    score = bip_metric(new_x_val, y_val, y_pred, rows_region)
    return score

In [9]:
def crossvalidation(df, rows_region, nfolds=8, n_estimators=50, n_jobs=1, verbose=False):
    '''Crossvalidation on the dataset `df` with `nfolds` folds.
    Split the dataset in N training-validation folds,
    trains and evaluates results for each of them,
    returns the mean of the error and metrics'''
    
    # convert date to datetime
    if df['Date'].dtype == np.dtype('O'):
        df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d')
        
    # add temporary columns to compute splits
    df['Year_CV'] = df['Date'].dt.year
    df['Month_CV'] = df['Date'].dt.month
    
    # get all year-month pairs
    l = sorted(list(set(df[['Year_CV', 'Month_CV']].itertuples(index=False))))
    
    # drop temporary columns
    df = df.drop(['Year_CV', 'Month_CV'], axis=1)
    
    # split dataset in folds
    kf = KFold(nfolds)
    indexes = []
    for train_months_idx, val_months_idx in kf.split(l):
        train_months = [l[i] for i in train_months_idx]
        val_months = [l[i] for i in val_months_idx]
        train_indexes = df.Date.apply(lambda d: (d.year, d.month) in train_months)
        val_indexes = df.Date.apply(lambda d: (d.year, d.month) in val_months)
        indexes.append((train_indexes, val_indexes))

    # iterate on the folds
    total_score = 0
    for train_indexes, val_indexes in indexes:
        # split
        df_train = df[train_indexes]
        df_validation = df[val_indexes]
        
        # store Date and ID
        val_date = pd.DataFrame(df_validation['Date'])
        val_store_id = df_validation['StoreID']
        
        # and drop them
        df_train = df_train.drop('Date', axis=1)
        df_validation = df_validation.drop('Date', axis=1)

        df_train = df_train.drop('StoreID', axis=1)
        df_validation = df_validation.drop('StoreID', axis=1)
        
        # train model
        X_train, y_train = get_x_y(df_train)
        model = train_model(X_train, y_train, n_estimators=n_estimators, n_jobs=n_jobs)
        
        # evaluate model
        X_val, y_val = get_x_y(df_validation)
#         months = pd.DatetimeIndex(val_date['Date']).month
        score = eval_model(X_val, y_val, model, val_store_id, val_date, rows_region)
        if verbose:
            print('Partial score: {:6.4f}'.format(score))
        total_score += score
        
    return total_score / nfolds

In [10]:
def crossvalidation_TS(df, rows_region, nfolds=8, n_estimators=50, n_jobs=1, verbose=False):
    '''Time-Series Crossvalidation on the dataset `df` with `nfolds` folds.
    Split the dataset in N training-validation folds,
    trains and evaluates results for each of them,
    returns the mean of the error and metrics'''
    
    # convert date to datetime
    if df['Date'].dtype == np.dtype('O'):
        df['Date'] = pd.to_datetime(df['Date'], format='%Y-%m-%d')
        
    # add temporary columns to compute splits
    df['Year_CV'] = df['Date'].dt.year
    df['Month_CV'] = df['Date'].dt.month
    
    # get all year-month pairs
    l = sorted(list(set(df[['Year_CV', 'Month_CV']].itertuples(index=False))))
    
    # drop temporary columns
    df = df.drop(['Year_CV', 'Month_CV'], axis=1)
    
    # split dataset in folds
    kf = TimeSeriesSplit(nfolds)
    indexes = []
    for train_months_idx, val_months_idx in kf.split(l):
        train_months = [l[i] for i in train_months_idx]
        val_months = [l[i] for i in val_months_idx]
        train_indexes = df.Date.apply(lambda d: (d.year, d.month) in train_months)
        val_indexes = df.Date.apply(lambda d: (d.year, d.month) in val_months)
        indexes.append((train_indexes, val_indexes))

    # iterate on the folds
    total_score = 0
    for train_indexes, val_indexes in indexes:
        # split
        df_train = df[train_indexes]
        df_validation = df[val_indexes]
        
        # store Date and ID
        val_date = pd.DataFrame(df_validation['Date'])
        val_store_id = df_validation['StoreID']
        
        # and drop them
        df_train = df_train.drop('Date', axis=1)
        df_validation = df_validation.drop('Date', axis=1)

        df_train = df_train.drop('StoreID', axis=1)
        df_validation = df_validation.drop('StoreID', axis=1)
        
        # train model
        X_train, y_train = get_x_y(df_train)
        model = train_model(X_train, y_train, n_estimators=n_estimators, n_jobs=n_jobs)
        
        # evaluate model
        X_val, y_val = get_x_y(df_validation)
#         months = pd.DatetimeIndex(val_date['Date']).month
        score = eval_model(X_val, y_val, model, val_store_id, val_date, rows_region)
        if verbose:
            print('Partial score: {:6.4f}'.format(score))
        total_score += score
        
    return total_score / nfolds

### Load dataset

In [11]:
# load preprocessed csv to dataframe
df = pd.read_csv('preprocessed_train.csv')

### Prepare dataset

In [12]:
# Save region for each index
rows_region = df['Region']

# Choose features

# df_train = df[selected_features]
# df_validation = df[selected_features]

In [13]:
# Sistemo i dati per regression tree
## StoreID
# df.drop('StoreID',axis=1) droppo dopo

## StoreType
df = pd.get_dummies(df, columns=['StoreType'], prefix='StoreType')

## AssortmentType
df = pd.get_dummies(df, columns=['AssortmentType'], prefix='AssortmentType')

## Region
# df = pd.get_dummies(df, columns=['Region'], prefix='Region')

## Events
# No-Events (NaN) are considered as sunny days, with lowest value (0) on the events scale
df['Events'] = df['Events'].fillna(0)
df=df.replace({'Rain':1, 'Thunderstorm':1, 'Fog':1, 'Snow': 2, 'Fog-Rain': 2, 'Rain-Thunderstorm': 2, 'Rain-Snow':2, 'Fog-Snow':2, 'Fog-Rain-Snow':3, 'Rain-Hail':3, 'Snow-Hail':3, 'Rain-Snow-Hail':3, 'Fog-Rain-Hail':3, 'Fog-Thunderstorm':3, 'Fog-Rain-Thunderstorm':4, 'Fog-Snow-Hail':4, 'Fog-Rain-Snow-Hail':4, 'Rain-Snow-Thunderstorm':4, 'Rain-Hail-Thunderstorm':4, 'Fog-Rain-Hail-Thunderstorm':4, 'Rain-Snow-Hail-Thunderstorm':4})

In [14]:
# Look for features with NaN values
null_cols = find_nan_features(df)
print('Features with NaN:')
for col in null_cols:
    print(col)
    
# drop all rows with NaN values
df = remove_nan_rows(df)

Features with NaN:
Max_VisibilityKm
Mean_VisibilityKm
Min_VisibilitykM
IsOpen_yesterday
IsOpen_tomorrow
IsHoliday_yesterday
IsHoliday_tomorrow
NumberOfSales_yesterday
NumberOfSales_lastweek
NumberOfSales_lastmonth
Average_numberOfSales_lastweek
Average_numberOfSales_lastmonth


### Cross-validation

In [15]:
score = crossvalidation(df, rows_region, nfolds=8, 
                        n_estimators=50, n_jobs=3, verbose=True)

print("Total score: {:6.4f}".format(score))

Partial score: 0.0312
Partial score: 0.0275
Partial score: 0.0518
Partial score: 0.0269
Partial score: 0.0312
Partial score: 0.0316
Partial score: 0.0413
Partial score: 0.0276
Total score: 0.0337


In [16]:
score_ts = crossvalidation_TS(df, rows_region, nfolds=8, 
                              n_estimators=50, n_jobs=3, verbose=True)

print("Total score: {:6.4f}".format(score_ts))

Partial score: 0.0741
Partial score: 0.0311
Partial score: 0.0275
Partial score: 0.0296
Partial score: 0.0268
Partial score: 0.0249
Partial score: 0.0495
Partial score: 0.0276
Total score: 0.0364


#### Cross-validation: Random Forest, 50 trees, 8 folds
    Partial score: 0.0215
    Partial score: 0.0203
    Partial score: 0.0356
    Partial score: 0.0227
    Partial score: 0.0263
    Partial score: 0.0204
    Partial score: 0.0243
    Partial score: 0.0316
    Total score: 0.0253

In [17]:
## questo notebook:
# senza StoreID
# con NumberOfSales reali
# con tutti i dati

## da fare:
# per store
# con i dati dello store

In [18]:
# TODO add MAE, MSE, errore per regione

##### 